contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1339 @@
+        );
+        token = _token;
+        keepStakingContract = _keepStakingContract;
+        nucypherStakingContract = _nucypherStakingContract;
+
+        keepFloatingPointDivisor = _keepVendingMachine.FLOATING_POINT_DIVISOR();
+        keepRatio = _keepVendingMachine.ratio();
+        nucypherFloatingPointDivisor = _nucypherVendingMachine
+            .FLOATING_POINT_DIVISOR();
+        nucypherRatio = _nucypherVendingMachine.ratio();
+    }
+
+    //
+    //
+    // Delegating a stake
+    //
+    //
+
+    /// @notice Creates a delegation with `msg.sender` owner with the given
+    ///         operator, beneficiary, and authorizer. Transfers the given
+    ///         amount of T to the staking contract.
+    /// @dev The owner of the delegation needs to have the amount approved to
+    ///      transfer to the staking contract.
+    function stake(
+        address _operator,
+        address payable _beneficiary,
+        address _authorizer,
+        uint96 _amount
+    ) external override {
+        require(_operator != address(0), "Operator must be specified");
+        require(_beneficiary != address(0), "Beneficiary must be specified");
+        require(_authorizer != address(0), "Authorizer must be specified");
+        OperatorInfo storage operator = operators[_operator];
+        (, uint256 createdAt, ) = keepStakingContract.getDelegationInfo(
+            _operator
+        );
+        require(
+            createdAt == 0 && operator.owner == address(0),
+            "Operator is already in use"
+        );
+        require(
+            _amount > minTStakeAmount,
+            "Amount to stake must be greater than minimum"
+        );
+        operator.owner = msg.sender;
+        operator.authorizer = _authorizer;
+        operator.beneficiary = _beneficiary;
+
+        operator.tStake = _amount;
+        /* solhint-disable-next-line not-rely-on-time */
+        operator.startTStakingTimestamp = block.timestamp;
+
+        emit TStaked(msg.sender, _operator);
+        emit OperatorStaked(
+            _operator,
+            operator.beneficiary,
+            operator.authorizer,
+            _amount
+        );
+        token.safeTransferFrom(msg.sender, address(this), _amount);
+    }
+
+    /// @notice Copies delegation from the legacy KEEP staking contract to T
+    ///         staking contract. No tokens are transferred. Caches the active
+    ///         stake amount from KEEP staking contract. Can be called by
+    ///         anyone.
+    function stakeKeep(address _operator) external override {
+        require(_operator != address(0), "Operator must be specified");
+        OperatorInfo storage operator = operators[_operator];
+
+        require(
+            operator.owner == address(0),
+            "Can't stake KEEP for this operator"
+        );
+
+        uint96 tAmount = getKeepAmountInT(_operator);
+
+        operator.keepInTStake = tAmount;
+        operator.owner = keepStakingContract.ownerOf(_operator);
```

Comment:
The owner can not be modified in Keep staking contract. It is either a liquid token delegation or a delegation from `TokenGrant`. `TokenGrant` grantee can be a simple grantee account address or [`ManagedGrant` contract](https://github.com/keep-network/keep-core/blob/main/solidity-v1/contracts/ManagedGrant.sol) that is proxying the calls and allows to switch grantee address.

To avoid resolving every single time `onlyOwnerOrOperator` is called, we could resolve real owner once in `stakeKeep` and allow to "refresh" the owner only if we are dealing with `ManagedGrant`. 

To resolve actual owner, we'd have to:
- Check if `keepStakingContract.ownerOf` is not `TokenGrant`. If not, use it as `operator.owner`.
- If `keepStakingContract.ownerOf` is `TokenGrant`, [check if `grantee`](https://github.com/keep-network/keep-core/blob/16c30709e185136e6fedeeca321a662dedf479c5/solidity-v1/contracts/TokenStakingEscrow.sol#L481-L487) is a [an instance of `ManagedGrant` contract](https://github.com/keep-network/keep-core/blob/16c30709e185136e6fedeeca321a662dedf479c5/solidity-v1/contracts/TokenStakingEscrow.sol#L407-L415). If not, use it as `operator.owner`.
-  If `keepStakingContract.ownerOf` is `TokenGrant` and `grantee` is an instance of `ManagerGrant` contract, take `grantee()` from there as `operator.owner` and expose `refreshKeepManagedGrantOwner` function allowing to repeat this procedure and update `operator.owner` in case `ManagedGrant.confirmGranteeReassignment` was called.

@eth-r thoughts on that?

Reply:
✔️